### PR TITLE
Feature attachments and detectors

### PR DESCRIPTION
**Pull request by jo3w4rd**
Detectors go into core, with examples in a DetectionExamples "module". This PR supersedes the PinchUtilities module, which still exists, but we should deprecate and remove after one more release.

Attachments go into Attachments module. Attachments still need an real example scene (the existing ones are just wip testing).

### CODE REVIEW

File: Assets/LeapMotion/Scenes/Leap_Hands_Demo_Desktop.unity [hunk 1]
```diff
@@ -186,6 +186,19 @@ Prefab:
       value: 1
       objectReference: {fileID: 0}
     - target: {fileID: 11406336, guid: b9c7aaf0d314cb847a3b2a20e893f1b6, type: 2}
+<<<<<<< HEAD
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scenes/Leap_Hands_Demo_Desktop.unity]: There is a merge conflict in this scene file

File: Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      if (IsActive) {
+        IsActive = false;
+        OnDeactivate.Invoke();
+      }
+      IsActive = false;
+    }
+  }
+
+  /** 
+  * Settings for handling pointing conditions
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs]: Does this belong in the base class? I don't think it would apply in all cases for extending detectors.

File: Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        handModel = gameObject.GetComponentInParent<IHandModel>();
+      }
+    }
+
+    private void OnEnable () {
+      StartCoroutine(fingerPointingWatcher());
+    }
+  
+    private void OnDisable () {
+      StopCoroutine(fingerPointingWatcher());
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs]: Calling fingerPointingWatcher() creates a new IEnumerator which is different than the one used to start the coroutine, so this method will do nothing.  You could save the IEnumerator or use StopAllCoroutines, or just start the coroutine in Start and never stop it.

File: Assets/LeapMotion/Scripts/Utils/Utils.cs [hunk 4]
```diff
@@ -29,5 +29,60 @@ public static class Utils {
...
+    /* Visual debugging utilities */
+    public static void DrawCircle(Vector3 center, 
+                           Vector3 normal, 
+                           float radius, 
+                           Color color, 
+                           int quality = 32, 
+                           float duration = 0, 
+                           bool depthTest = true) {
+      Vector3 planeA = Vector3.Slerp(normal, -normal, 0.5f);
+      Vector3 planeB = Vector3.Cross(normal, planeA).normalized;
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/Utils/Utils.cs]: This variable is not used.

File: Assets/LeapMotionModules/Attachments/Scripts/Transition.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        localPosition.z = ZPosition.Evaluate(interpolationPoint) * RelativeOnPosition.z;
+        transform.localPosition = localPosition;
+      }
+      if(AnimateRotation){
+        Quaternion transitionRotation = Quaternion.Euler(transform.localRotation.x + XRotation.Evaluate(interpolationPoint) * RelativeOnRotation.x,
+                                                         transform.localRotation.y + YRotation.Evaluate(interpolationPoint) * RelativeOnRotation.y,
+                                                         transform.localRotation.z + ZRotation.Evaluate(interpolationPoint) * RelativeOnRotation.z);
+        transform.localRotation = transitionRotation;
+      }
+      if (AnimateRotation && false) {
```
Comment (Amarcolina) [inline Assets/LeapMotionModules/Attachments/Scripts/Transition.cs]: Impossible condition

**Comment by Amarcolina** [comment]
The pinch utility should probably be removed so that things don't get mixed up.  Currently we have a duplicate component for everything that was in that module, and DetectionExample scenes are referencing the wrong script in several places.

File: Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+    * Invoked when this detector deactivates.
+    * Subclasses must call this function when the detector's conditions change from true to false.
+    */
+    public virtual void Deactivate(){
+      if (IsActive) {
+        IsActive = false;
+        OnDeactivate.Invoke();
+      }
+      IsActive = false;
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs]: This statement is unneeded.

File: Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+    * Invoked when this detector activates.
+    * Subclasses must call this function when the detector's conditions become true.
+    */
+    public virtual void Activate(){
+      if (!IsActive) {
+        IsActive = true;
+        OnActivate.Invoke();
+      }
+      IsActive = true;
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs]: Also unneeded.

File: Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
+﻿using UnityEngine;
+using UnityEngine.Events;
+using System.Collections;
+using Leap;
+
+namespace Leap.Unity {
+
+  public class Detector : MonoBehaviour {
+    public bool IsActive{ get{ return _isActive;} private set { _isActive = value;}}
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs]: Having a private setter in this case isn't needed since there already is a way to set the variable privately.  Having a setter also might open the door for potential bugs if additional code is later added to the setter unintentionally.

File: Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    public void RemoveDetector(Detector detector){
+      detector.OnActivate.RemoveListener(CheckDetectors);
+      detector.OnDeactivate.RemoveListener(CheckDetectors);
+      Detectors.Remove(detector);
+    }
+
+    public void AddAllDetectors(){
+      Detector[] detectors = GetComponents<Detector>();
+      for(int g = 0; g < detectors.Length; g++){
+        if ( detectors[g].GetInstanceID() != this.GetInstanceID() && detectors[g].enabled) {
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs]: any reason you are using instanceID instead of regular equality?

File: Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+          AddDetector(detectors[g]);
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs]: We should check if the detector is already added, since a user might easily add a sibling detector component to the list.

Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs]: Ah I see, that is done inside the AddDetector method, nevermind!

File: Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs [hunk 11]
```diff
@@ -0,0 +1,79 @@
+﻿using UnityEngine;
+using UnityEngine.Events;
+using System.Collections.Generic;
+
+namespace Leap.Unity {
+
+  public class DetectorLogicGate : Detector {
+    public List<Detector> Detectors;
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs]: This list should not be public, since modification would cause incorrect behaviour.

File: Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    public bool AddAllDetectorsOnAwake = true;
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs]: Maybe rename to AddSiblingDetectorsOnAwake instead?  It's not immediately obvious what 'All' refers to.

File: Assets/LeapMotion/Scripts/DetectionUtilities/ExtendedFingerDetector.cs [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+  
+    public PointingState Thumb = PointingState.Either;
+    public PointingState Index = PointingState.Either;
+    public PointingState Middle = PointingState.Either;
+    public PointingState Ring = PointingState.Either;
+    public PointingState Pinky = PointingState.Either;
+  
+    void Start () {
+      if(handModel == null){
+        handModel = gameObject.GetComponentInParent<IHandModel>();
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/ExtendedFingerDetector.cs]: Start happens after OnEnable, so if the hand model is not assigned, the coroutine will give an error since it tries to access handModel which might be null.

File: Assets/LeapMotion/Scripts/DetectionUtilities/ExtendedFingerDetector.cs [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+          }
+        } else if(IsActive){
+          Deactivate();
+        }
+        yield return new WaitForSeconds(Period);
+      }
+    }
+  
+    private bool matchFingerState(Finger finger, int ordinal){
+      switch(ordinal){
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/ExtendedFingerDetector.cs]: This could be simplified so that the switch only selects which PointingState to use, and then the conditional logic doesn't have to be repeated.

File: Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+      while(true){
+        if(handModel != null){
+          hand = handModel.GetLeapHand();
+          if(hand != null){
+            targetDirection = selectedDirection(hand.Fingers[selectedFinger].TipPosition.ToVector3());
+            fingerDirection = hand.Fingers[selectedFinger].Direction.ToVector3();
+            float angleTo = Vector3.Angle(fingerDirection, targetDirection);
+            if(handModel.IsTracked && angleTo <= OnAngle){
+              Activate();
+            } else if (!handModel.IsTracked || angleTo >= OffAngle) {
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs]: the elseif is not needed, just an else would have the same effect.

File: Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+        case PointingType.RelativeToWorld:
+          return PointingDirection;
+        case PointingType.AtTarget:
+          return TargetObject.position - tipPosition;
+        default:
+          return PointingDirection;
+      }
+    }
+
+    private int selectedFingerOrdinal(){
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs]: These are hardcoded indexes into the finger array.  Is that order ensured to be the same?  And if so, we should provide a way to use the enum to index the array within our Finger class instead of a strange switch statement over here.

File: Assets/LeapMotion/Scripts/DetectionUtilities/PalmDirectionDetector.cs [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+      Vector3 normal;
+      while(true){
+        if(handModel != null){
+          hand = handModel.GetLeapHand();
+          if(hand != null){
+            normal = hand.PalmNormal.ToVector3();
+            float angleTo = Vector3.Angle(normal, selectedDirection(hand.PalmPosition.ToVector3()));
+            if(angleTo <= OnAngle){
+              Activate();
+            } else if(angleTo > OffAngle) {
```
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/PalmDirectionDetector.cs]: Unneeded elseif, else is fine.

File: Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs (hunk 8 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs]: removed.

File: Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs (hunk 7 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs]: removed

File: Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs (hunk 6 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs]: removed.

File: Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs (hunk 2 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/Detector.cs]: Moved to palm pointing detector.

File: Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs (hunk 11 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs]: privatized.

File: Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs (hunk 12 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs]: renamed

File: Assets/LeapMotion/Scripts/DetectionUtilities/ExtendedFingerDetector.cs (hunk 13 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/ExtendedFingerDetector.cs]: Changed to Awake.

File: Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs (hunk 3 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs]: Odd, I thought I had switched them all over to saving a reference to the IEnumerator... Well, they do now.

File: Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs (hunk 15 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs]: I don't think so -- I don't want to deactivate until angleTo becomes grater than OffAngle,  not just when angleTo is no longer less than or equal to OnAngle.

File: Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs (hunk 16 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs]: I'd still have to write a function to get the ordinal because the FingerType has a TYPE_UNKNOWN, which I can't use.

File: Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs (hunk 15 above)
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs]: Ah yes of course, sorry missed the OnAngle being different than OffAngle :)

File: Assets/LeapMotion/Scripts/DetectionUtilities/PalmDirectionDetector.cs (hunk 17 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/PalmDirectionDetector.cs]: As above, I think it is needed.

File: Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs (hunk 16 above)
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs]: But currently TYPE_UNKNOWN can be used by this switch statement, and defaults to the index finger.

File: Assets/LeapMotion/Scripts/DetectionUtilities/PalmDirectionDetector.cs (hunk 17 above)
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/PalmDirectionDetector.cs]: Yup, you are very correct!

File: Assets/LeapMotion/Scripts/DetectionUtilities/ExtendedFingerDetector.cs (hunk 14 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/ExtendedFingerDetector.cs]: Simplified.

File: Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs (hunk 16 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs]: Maybe I'm misunderstanding your suggestion... I can't use the values of the selected enum member to index the array even though they match up with the finger array position, because TYPE_UNKNOWN is -1. Were you suggesting a change to the Finger class?

File: Assets/LeapMotion/Scenes/Leap_Hands_Demo_Desktop.unity (hunk 1 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scenes/Leap_Hands_Demo_Desktop.unity]: deconflicted.

File: Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs (hunk 9 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/DetectorLogicGate.cs]: Probably no good reason. Changed.

File: Assets/LeapMotion/Scripts/Utils/Utils.cs (hunk 4 above)
Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/Utils/Utils.cs]: ANd removed.

File: Assets/LeapMotionModules/Attachments/Scripts/Transition.cs (hunk 5 above)
Comment (jo3w4rd) [inline Assets/LeapMotionModules/Attachments/Scripts/Transition.cs]: Removed rejected alternative calculation.

File: Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs (hunk 16 above)
Comment (Amarcolina) [inline Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs]: Yeah I am proposing that we add a method to Hand that takes in a finger type and returns the finger of that type (or null / error if you pass in type unknown).  The hand class could use the enum value as an index, just with some extra checks for unknown enum types.  That way this functionality (which is quite useful in all cases!) can be used all over the place.

Comment (jo3w4rd) [inline Assets/LeapMotion/Scripts/DetectionUtilities/FingerDirectionDetector.cs]: That will have to be a different PR.

**Comment by jo3w4rd** [comment]
@Amarcolina I think I've addressed all your current comments.

**Comment by Amarcolina** [comment]
Everything works great!
